packages/menu/src/useMenu.ts

```diff
@@ -0,0 +1,644 @@
+  }, [onChange]);
+
+  /**
+   * Effects
+   */
+
+  /**
+   * Sets a state value to ensure 1 render happens after
+   * opening so a menu item can be focused.
+   */
+  useEffect(() => {
+    setMenuVisible(controlledIsExpanded);
+  }, [controlledIsExpanded, menuRef, environment]);
+
+  /**
+   * Respond to clicks outside the  open menu
+   */
+  useEffect(() => {
+    const doc = environment || document;
+
+    if (controlledIsExpanded) {
+      doc.addEventListener('click', handleMenuBlur, true);
+    } else if (!controlledIsExpanded) {
+      doc.removeEventListener('click', handleMenuBlur, true);
+    }
+
+    return () => {
+      doc.removeEventListener('click', handleMenuBlur, true);
+    };
+  }, [controlledIsExpanded, handleMenuBlur, environment]);
+
+  /**
+   * Return focus to trigger when menu closes
+   */
+  useEffect(() => {
+    if (controlledIsExpanded && !returnFocusRef.current) {
+      returnFocusRef.current = triggerRef.current;
+    }
+
+    if (!controlledIsExpanded && returnFocusRef.current) {
+      setTimeout(() => {
+        returnFocusRef.current!.focus();
+        returnFocusRef.current = null;
+      });
+    }
+  }, [triggerRef, controlledIsExpanded]);
+
+  /**
+   * Focus initial item when menu opens
+   */
+  useEffect(() => {
+    if (state.focusOnOpen && menuVisible && controlledFocusedValue && controlledIsExpanded) {
+      const ref = itemRefs[controlledFocusedValue]?.current;
+
+      ref && ref.focus();
+    }
+  }, [menuVisible, itemRefs, controlledFocusedValue, state.focusOnOpen, controlledIsExpanded]);
+
+  /**
+   * Prop getters
+   */
+
+  const getTriggerProps = useCallback<IUseMenuReturnValue<T, L>['getTriggerProps']>(
+    ({ onClick, onKeyDown, type = 'button', role = 'button', disabled, ...other } = {}) => ({
+      'data-garden-container-id': 'containers.menu.trigger',
+      'data-garden-container-version': PACKAGE_VERSION,
+      ref: triggerRef,
+      id: triggerId,
+      'aria-expanded': state.isExpanded,
+      'aria-haspopup': true,
+      disabled,
+      tabIndex: disabled ? -1 : 0,
+      type: type === null ? undefined : type,
+      role: role === null ? undefined : role,
+      ...other,
+      onKeyDown: composeEventHandlers(onKeyDown, handleTriggerKeyDown),
+      onClick: composeEventHandlers(onClick, handleTriggerClick)
+    }),
+    [triggerRef, state.isExpanded, handleTriggerClick, handleTriggerKeyDown, triggerId]
+  );
+
+  const getMenuProps = useCallback<IUseMenuReturnValue<T, L>['getMenuProps']>(
+    ({ role = 'menu', onKeyDown, onMouseLeave, ...other } = {}) => ({
+      ...getGroupProps({
+        onKeyDown: composeEventHandlers(onKeyDown, handleMenuKeyDown),
+        onMouseLeave: composeEventHandlers(onMouseLeave, handleMenuMouseLeave)
+      }),
+      'data-garden-container-id': 'containers.menu',
+      'data-garden-container-version': PACKAGE_VERSION,
+      'aria-labelledby': triggerId,
+      role: role === null ? undefined : role,
+      ref: menuRef,
```

Comment:
Hmm we might need to have a deeper conversation about this, because as of the recent `useSelection`/`useTabs` update, both those hooks return `ref`s in their element/tab getters, so that's what I based this approach on 😬

Reply:
Yeah, I missed that on review. It would be good to be consistent across the container codebase. I'm not particularly fond of a consumer providing a ref, we loop it back, and then having a `mergeRefs` happen between two identical refs.

Reply:
How are you leaning about this after the weekend? Should we have a deeper discussion about it?

I’m kinda split. I did some brief research on popular no-UI prop collection hooks ([react-table](https://github.com/TanStack/table/blob/main/packages/react-table/src/index.tsx), [reach-ui/tooltip](https://github.com/reach/reach-ui/blob/dev/packages/tooltip/src/reach-tooltip.tsx#L362), [downshift useCombobox](https://github.com/downshift-js/downshift/blob/master/src/hooks/useCombobox/index.js#L477)) and there was a somewhat decent mix of `ref` vs no `ref` required in args and/or in the return value.

Reply:
I think the prevailing pattern in Garden's `react-containers` is to receive refs and not give them back in the return getters. I'd like to follow suit here, and make the future `useSelection` related adjustments to stick with that approach.

Reply:
Per conversation, going to leave the ref for `getItemProps` in the return value. :+1: 

`useMenu` will still receive `triggerRef` and `menuRef`.